Delayed unused IP address cleanup.

veshij opened the pull request:
**What type of PR is this?**: feature
**Which issue does this PR fix**: https://github.com/aws/amazon-vpc-cni-k8s/issues/1700
**What does this PR do / Why do we need it**: TODO
**Testing done on this change**: Y
**Automation added to e2e**: No
**Will this break upgrades or downgrades. Has updating a running cluster been tested?**: No/No
**Does this change require updates to the CNI daemonset config files to work?**: No
**Does this PR introduce any user-facing change?**: N
```release-note
Free IP addresses are released with a 10-minute delay to avoid unnecessary churn on loaded nodes.
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

cgchinmay commented:
@veshij Could you update your branch. Just started reviewing it. Thanks

jayanthvn commented on pkg/ipamd/datastore/data_store.go:
```diff
@@ -34,6 +34,9 @@ const (
 	addressCoolingPeriod = 30 * time.Second
 
+	// addressDelayedReleasePeriod is used to release unused IP after a period of inactivity.
+	addressDelayedReleasePeriod = 10 * time.Minute
```
Still reviewing it, but on a high level this should be configurable.

veshij replied:
Sure, I can make it configurable.
afair there was a discussion in related issue that we don't want to introduce new env variables. Is it ok to expose it via env var with some safe default value?

jayanthvn commented:
Thanks for the changes. We will provide review comments and given this is in the critical IP allocation path, we would also like to test it on our cluster before merging. Currently we are working on a release and holidays are coming up so will keep you updated once we start with the testing for this PR.

achevuru commented:
I have a general comment/observation. Why can't we achieve this with WARM and MIN IP targets? Understand we want to avoid unnecessary assign/detach IP calls in a high churn env but a user can greatly minimize that by configuring WARM and MIN IP values based on their expected churn/use case, isn't it? Now, holding a free IP for 10 mins by default might be too conservative as we're taking these IPs from VPC IP pool and they are not just unique at cluster level.

veshij commented:
> I have a general comment/observation. Why can't we achieve this with WARM and MIN IP targets? Understand we want to avoid unnecessary assign/detach IP calls in a high churn env but a user can greatly minimize that by configuring WARM and MIN IP values based on their expected churn/use case, isn't it? Now, holding a free IP for 10 mins by default might be too conservative as we're taking these IPs from VPC IP pool and they are not just unique at cluster level.

* cooldown addresses need to be accounted as "busy" since those can't be assigned to pods by request. If we are making this change CNI will start allocating additional addresses for short period of time while recently released addresses are in cooldown unnecessary increasing  address churn and api requests.
* WARM/MIN ip target logic works perfectly fine in more or less stable environments when number of pods don't change over time and  targets are chosen correctly. Otherwise there are two options: either nodes are "over-provisioned" and many addresses are not used in the cluster or we have to wait each time pod is launched to allocate an address.

10 minutes limit can be safely decreased, it's just a timeout which we found to work great in our environment.
I can make it low by default (2*cooldown?) and add an environment variable to configure it.

veshij commented:
@jayanthvn @achevuru what do you think?

jayanthvn commented:
@veshij - We both are on PTO until next week end. We will get back to you on the week of 17th. Sorry for the delay.

Reg the environment variable, yes it would be good to have. If environment variable is "not configured" or "0" then we can fall back to existing design where WARM targets decide the number of IPs in the pool else if set then released IPs will stay in pool for "cooldown period + delayed release period".

veshij commented:
Sorry for the quite long delay.

> Reg the environment variable, yes it would be good to have. If environment variable is "not configured" or "0" then we can fall back to existing design where WARM targets decide the number of IPs in the pool else if set then released IPs will stay in pool for "cooldown period + delayed release period".
@jayanthvn I introduced the env variable, updated code to use it.
I defaulted it to 60 seconds (so address will spend 30s in "cooldown" and another 30s in delayed release state). In our environment even conservative 30s gives us quite good address reuse.
Please let me know if you'd prefer to default to 0.
I will update corresponding docs if this approach makes sense to you.
Thank you!

jayanthvn commented:
@veshij - Will keep it open.

jayanthvn commented:
/not stale. We need to continue discussion on this.
